velox/functions/prestosql/ArrayPosition.cpp

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <deque>
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/vector/DecodedVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <TypeKind kind>
+void applyTypedFirstMatch(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
```

Comment:
That's right. But I was worrying if I move the code that obtain `elementsDecoded` from `arrayDecoded` into this function, it may cause code duplication because I need to add the same code to `applyComplexTypeFristMatch`, `applyTypedWithInstance`, and `applyComplexTypeWithInstance`.

Reply:
Well we seem to be rewriting baseArray, rawSizes etc.. so not sure why elementsDecoded cant also be there.

Reply:
Yeah, this is true... OK, I'll move the computation of elementDecoded into this function. Thank you for the suggestion!

Reply:
So I didn't move the computation of `elementDecoded` into the function after an attempt. The reason is that I found this decoding computation requires a `context` variable from the caller of `applyTypedFirstMatch`, so if I move the decoding computation into this function, I'll have to add `context` as a parameter and thus not able to shorten the parameter list anyway.

Reply:
So I didn't move the decoding of elements vector into this function after an attempt. I found that the decoding requires a `context` variable from the caller of this function, so if I move the decoding inside to eliminate the `elementsDecoded` parameter, I'll have a add another parameter.